harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -20,55 +19,97 @@
     Reducer,
     TorchData,
     _callback,
-    _Data,
     _reduce_metrics,
     data_length,
-    to_device,
 )
 from determined_common import check
 
-# Apex is included only for GPU trials.
-try:
-    import apex
-except ImportError:
-    if torch.cuda.is_available():
-        logging.warning("Failed to import apex.")
-    pass
-
-
-class _WarningLogs(enum.Enum):
-    FAILED_MOVING_TO_DEVICE = 1
-
 
 class PyTorchTrialController(det.LoopTrialController):
     def __init__(self, trial_inst: det.Trial, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
 
         check.is_instance(trial_inst, PyTorchTrial, "PyTorchTrialController needs an PyTorchTrial")
         self.trial = cast(PyTorchTrial, trial_inst)
-        self._check_evaluate_implementation()
+        self.context = cast(PyTorchTrialContext, self.context)
+        self.context.callbacks = self.trial.build_callbacks()
+
+        # TODO(DET-3262): remove this backward compatibility of old interface.
+        if (
+            util.is_overridden(self.trial.build_model, PyTorchTrial)
+            or util.is_overridden(self.trial.optimizer, PyTorchTrial)
+            or util.is_overridden(self.trial.create_lr_scheduler, PyTorchTrial)
+        ):
+            check.true(
+                util.is_overridden(self.trial.build_model, PyTorchTrial)
+                and util.is_overridden(self.trial.optimizer, PyTorchTrial),
+                "Both build_model() and optimizer() must be defined "
+                "if any of build_model(), optimizer(), and create_lr_scheduler() are defined. "
+                "If you want to use the new interface, you should instead instantiate your model, "
+                "optimizers, and LR schedulers in __init__ and call context.backward(loss) "
+                "and context.step_optimizer(optimizer) in train_batch.",
+            )
+
+            model = self.context._Model(self.trial.build_model())
+            optim = self.context._Optimizer(self.trial.optimizer(model))
+
+            lr_scheduler = self.trial.create_lr_scheduler(optim)
+            if lr_scheduler is not None:
+                self.context.lr_schedulers.append(lr_scheduler)
+
+            if det.ExperimentConfig(self.context.get_experiment_config()).mixed_precision_enabled():
+                self.context._configure_amp(
+                    model=model,
+                    optimizers=optim,
+                    opt_level=self.context.get_experiment_config()
+                    .get("optimizations", {})
+                    .get("mixed_precision", "O0"),
+                )
+
+            train_batch = self.trial.train_batch
+
+            def new_train_batch(
+                batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
```

Comment:
question: at some point we are going to take the `model` argument out, right?  When is the right time to do that?

Reply:
Since we only have one main model for each trial class, I think it might make sense to keep this argument. Don't see any additional benefits to make changes so I'd rather not make any changes

Reply:
Update: will remove model in the future since we are introducing multiple models